File: app/src/main/java/fup/prototype/robprototype/details/DetailFragment.java
package fup.prototype.robprototype.details;

import android.arch.lifecycle.ViewModelProviders;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;

import com.rxdroid.api.error.RequestError;
import com.rxdroid.repository.model.User;

import javax.inject.Inject;

import fup.prototype.robprototype.R;
import fup.prototype.robprototype.databinding.FragmentDetailsBinding;
import fup.prototype.robprototype.util.DialogUtils;
import fup.prototype.robprototype.view.LiveDataViewModelFactory;
import fup.prototype.robprototype.view.base.fragments.DataFragment;

public class DetailFragment extends DataFragment<FragmentDetailsBinding, DetailViewModel> {

    private static final String KEY_USER = "keyUser";
    private static final String KEY_USER_SAVE = "keyUserSave";

    @Inject
    protected LiveDataViewModelFactory liveDataViewModelFactory;

    public static DetailFragment newInstance(@NonNull final Bundle bundle) {
        final DetailFragment detailFragment = new DetailFragment();
        detailFragment.setArguments(bundle);
        return detailFragment;
    }

    public static Bundle createBundle(@NonNull final User user) {
        final Bundle bundle = new Bundle();
        bundle.putSerializable(KEY_USER, user);
        return bundle;
    }

    public DetailFragment() {
        // empty
    }

    @Override
    public DetailViewModel createViewModel() {
        return ViewModelProviders.of(this, liveDataViewModelFactory).get(DetailViewModel.class);
    }

    @Override
    public void initBinding(final FragmentDetailsBinding binding) {
        binding.setViewModel(getViewModel());
        final User user = (User) getArguments().getSerializable(KEY_USER);
        getViewModel().setUser(user);
        setupRepositoryAdapter();
    }

    private void setupRepositoryAdapter() {
        final RecyclerView recyclerView = getViewBinding().recyclerView;
        final RepoAdapter repoAdapter = new RepoAdapter();
        final LinearLayoutManager linearLayoutManager = new LinearLayoutManager(getContext(), LinearLayoutManager.VERTICAL, false);
        recyclerView.setLayoutManager(linearLayoutManager);
        recyclerView.setAdapter(repoAdapter);
    }

    @Override
    public int getLayoutId() {
        return R.layout.fragment_details;
    }

    @Override
    public void addViewListener() {
        //nothing
    }

    @Override
    public void onResume() {
        super.onResume();
        getViewModel().loadOrShowData();
    }

    @Override
    protected AlertDialog createErrorDialog(@NonNull final RequestError requestError) {
        return DialogUtils.createOkCancelDialog(getContext(), "Möp", "A wild error occurred", "Ok", "Fuck it", null, null);
    }

    @Override
    public String getKey() {
        return DetailFragment.class.getSimpleName();
    }
}

File: app/src/main/java/fup/prototype/robprototype/details/DetailFragment.kt
package fup.prototype.robprototype.details

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.os.Bundle
import android.support.v7.app.AlertDialog
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import com.rxdroid.api.error.RequestError
import com.rxdroid.repository.model.User
import fup.prototype.robprototype.BR
import fup.prototype.robprototype.R
import fup.prototype.robprototype.databinding.FragmentDetailsBinding
import fup.prototype.robprototype.util.DialogUtils
import fup.prototype.robprototype.view.ViewModelFactory
import fup.prototype.robprototype.view.base.fragments.DataFragment
import javax.inject.Inject

class DetailFragment : DataFragment<FragmentDetailsBinding, DetailViewModel>() {

    @Inject
    lateinit var viewModelFactory: ViewModelFactory

    private var repositoryAdapter = RepositoryAdapter()

    companion object {
        fun newInstance(user: User): DetailFragment {
            val bundle = Bundle()
            bundle.putSerializable(DetailActivity.DetailConstants.KEY_USER, user)
            val detailFragment = DetailFragment()
            detailFragment.arguments = bundle
            return detailFragment
        }
    }

    override fun initBinding(binding: FragmentDetailsBinding?) {
        binding?.setVariable(BR.viewModel, getViewModel())
        binding?.setLifecycleOwner(this)
        setupRepositoryAdapter()
    }

    private fun setupRepositoryAdapter() {
        val recyclerView: RecyclerView? = getViewBinding()?.recyclerView
        recyclerView?.adapter = repositoryAdapter
        val linearLayoutManager = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
        recyclerView?.layoutManager = linearLayoutManager
    }

    override fun createViewModel(): DetailViewModel? {
        return ViewModelProviders.of(this, viewModelFactory).get(DetailViewModel::class.java)
    }

    override fun getLayoutId(): Int {
        return R.layout.fragment_details
    }

    override fun onResume() {
        super.onResume()
        val user: User = arguments.getSerializable(DetailActivity.DetailConstants.KEY_USER) as User
        getViewModel()?.loadReposForUser(user)
    }

    override fun applyLiveDataObserver() {
        getViewModel()?.getItems()?.observe(this, Observer { items ->
            kotlin.run {
                if (items != null && !items.isEmpty()) {
                    repositoryAdapter.clearAndAddItems(items)
                }
            }
        })
    }

    override fun addViewListener() {
        //nothing
    }

    override fun createErrorDialog(requestError: RequestError): AlertDialog {
        return DialogUtils.createOkCancelDialog(context, "Möp", "A wild error occurred", "Ok", "Fuck it", null, null)

    }

    override fun getKey(): String {
        return DetailFragment.toString()
    }

}